gateway-service/src/main/java/org/zowe/apiml/gateway/ribbon/loadbalancer/predicate/AuthenticationBasedPredicate.java

```diff
@@ -0,0 +1,86 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.zowe.apiml.gateway.cache.LoadBalancerCache;
+import org.zowe.apiml.gateway.ribbon.loadbalancer.LoadBalancingContext;
+import org.zowe.apiml.gateway.ribbon.loadbalancer.RequestAwarePredicate;
+import org.zowe.apiml.gateway.ribbon.loadbalancer.model.LoadBalancerCacheRecord;
+import org.zowe.apiml.gateway.security.service.HttpAuthenticationService;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Optional;
+
+import static org.springframework.cloud.netflix.zuul.filters.support.FilterConstants.SERVICE_ID_KEY;
+
+/**
+ * Based on the authentication information decide which instance should be used.
+ * If the user is authenticated and already has routing information stored in cache, use the information.
+ * <p>
+ * There is also terrible overhead as this happens for all instance ids.
+ */
+@RequiredArgsConstructor
+public class AuthenticationBasedPredicate extends RequestAwarePredicate {
+    private final HttpAuthenticationService authenticationService;
+    private final LoadBalancerCache cache;
+
+    @Value("${instance.metadata.apiml.lb.cacheRecordExpirationTimeInHours:8}")
+    int expirationTime;
+
+    @Override
+    public boolean apply(LoadBalancingContext context, DiscoveryEnabledServer server) {
+        RequestContext requestContext = context.getRequestContext();
+        String serviceId = (String) requestContext.get(SERVICE_ID_KEY);
+        if (serviceId == null) {
+            // This should never happen
+            return true;
+        }
+
+        Optional<String> authenticatedUser = authenticationService.getAuthenticatedUser(requestContext.getRequest());
+
+        if (!authenticatedUser.isPresent()) {
+            // Allow selection of any instance.
+            return true;
+        }
+
+        String username = authenticatedUser.get();
+        LoadBalancerCacheRecord loadBalancerCacheRecord = cache.retrieve(username, serviceId);
+        if (loadBalancerCacheRecord != null && loadBalancerCacheRecord.getInstanceId() != null) {
+            long creationTime = loadBalancerCacheRecord.getCreationTime();
+            if (isTooOld(creationTime)) {
+                cache.delete(username, serviceId);
+                return true;
+            } else {
+                return server.getInstanceInfo().getInstanceId().equalsIgnoreCase(loadBalancerCacheRecord.getInstanceId());
+            }
+        } else {
+            // There is no preference for given user
+            return true;
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "AuthenticationBasedPredicate (USERNAME)";
+    }
+
+    private boolean isTooOld(long cachedDate) {
+        Calendar calendar = Calendar.getInstance();
+        Date now = new Date();
+        calendar.setTime(new Date(cachedDate));
+        calendar.add(Calendar.HOUR, expirationTime);
+        return now.after(calendar.getTime());
+    }
```

Comment:
The old Date and Calendar api makes this a cumbersome task. The newer time impl. makes this one liner: Duration.between(LocalDateTime1, LocalDateTime2).inSeconds(); Not a big deal but you might consider learning this api.

Reply:
@taban03 This may be relevant to you.

Reply:
Sounds way better, I'll take a look on this api